src/napari_graph/undirected_graph.py

```diff
@@ -0,0 +1,306 @@
+        raise ValueError("Edge buffer is full.")
+
+    elif empty_idx < 0:
+        raise ValueError("Invalid empty index.")
+
+    next_edge = node2edge[src_node]
+    node2edge[src_node] = empty_idx
+
+    buffer_index = empty_idx * _UN_EDGE_SIZE
+    next_empty = buffer[buffer_index + _LL_UN_EDGE_POS]
+
+    buffer[buffer_index] = src_node
+    buffer[buffer_index + 1] = tgt_node
+    buffer[buffer_index + _LL_UN_EDGE_POS] = next_edge
+
+    return next_empty
+
+
+@njit
+def _add_undirected_edges(
+    buffer: np.ndarray,
+    edges: np.ndarray,
+    empty_idx: int,
+    n_edges: int,
+    node2edge: np.ndarray,
+) -> Tuple[int, int]:
+    """Add an array of edges into the `buffer`.
+
+    Edges are duplicated so both directions are available for fast graph
+    transversal.
+    """
+    size = edges.shape[0]
+    for i in range(size):
+
+        # adding (u, v)
+        empty_idx = _add_undirected_edge(
+            buffer, node2edge, empty_idx, edges[i, 0], edges[i, 1]
+        )
+        # adding (v, u)
+        empty_idx = _add_undirected_edge(
+            buffer, node2edge, empty_idx, edges[i, 1], edges[i, 0]
+        )
+
+        n_edges += 1
+
+    return empty_idx, n_edges
+
+
+@njit
+def _remove_undirected_edge(
+    src_node: int,
+    tgt_node: int,
+    empty_idx: int,
+    edges_buffer: np.ndarray,
+    node2edges: np.ndarray,
+) -> int:
+    """Remove a single edge (and its duplicated sibling edge) from the buffer.
+
+    NOTE: Edges are removed such that empty pairs are consecutive in memory.
+    """
+    empty_idx = _remove_edge(
+        tgt_node,
+        src_node,
+        empty_idx,
+        edges_buffer,
+        node2edges,
+        _UN_EDGE_SIZE,
+        _LL_UN_EDGE_POS,
+    )
+
+    empty_idx = _remove_edge(
+        src_node,
+        tgt_node,
+        empty_idx,
+        edges_buffer,
+        node2edges,
+        _UN_EDGE_SIZE,
+        _LL_UN_EDGE_POS,
+    )
+
+    return empty_idx
+
+
+@njit
+def _remove_undirected_edges(
+    edges: np.ndarray,
+    empty_idx: int,
+    edges_buffer: np.ndarray,
+    node2edges: np.ndarray,
+) -> int:
+    """Remove an array of edges (and their duplicated siblings) from buffer."""
+
+    for i in range(edges.shape[0]):
+        empty_idx = _remove_undirected_edge(
+            edges[i, 0], edges[i, 1], empty_idx, edges_buffer, node2edges
+        )
+    return empty_idx
+
+
+@njit
+def _remove_incident_undirected_edges(
+    node: int,
+    empty_idx: int,
+    n_edges: int,
+    edges_buffer: np.ndarray,
+    node2edges: np.ndarray,
+) -> Tuple[int, int]:
+    """Removes every edges that contains `node_idx`.
+
+    NOTE: Edges are removed such that empty pairs are consecutive in memory.
+    """
+
+    # the edges are removed such that the empty edges linked list contains
+    # two positions adjacent in memory so we can serialize the edges using
+    # numpy vectorization
+    idx = node2edges[node]
+
+    while idx != _EDGE_EMPTY_PTR:
```

Comment:
As above — avoid bare while loops, make sure it's a for-loop, and maybe raise an error if you reach the end